File: 0044-wildcard-matching/0044-wildcard-matching.java
public class Solution {
    public boolean isMatch(String s, String p) {
        int sIndex = 0, pIndex = 0;
        int starIndex = -1, sTempIndex = -1;

        while (sIndex < s.length()) {
            if (pIndex < p.length() && (s.charAt(sIndex) == p.charAt(pIndex) || p.charAt(pIndex) == '?')) {
                sIndex++;
                pIndex++;
            } else if (pIndex < p.length() && p.charAt(pIndex) == '*') {
                starIndex = pIndex;
                sTempIndex = sIndex;
                pIndex++;
            } else if (starIndex != -1) {
                pIndex = starIndex + 1;
                sTempIndex++;
                sIndex = sTempIndex;
            } else {
                return false;
            }
        }

        while (pIndex < p.length() && p.charAt(pIndex) == '*') {
            pIndex++;
        }

        return pIndex == p.length();
    }
}

File: 0044-wildcard-matching/0044-wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        sIndex, pIndex = 0, 0
        starIndex, sTempIndex = -1, -1

        while sIndex < len(s):
            if pIndex < len(p) and (s[sIndex] == p[pIndex] or p[pIndex] == '?'):
                sIndex += 1
                pIndex += 1
            elif pIndex < len(p) and p[pIndex] == '*':
                starIndex = pIndex
                sTempIndex = sIndex
                pIndex += 1
            elif starIndex != -1:
                pIndex = starIndex + 1
                sTempIndex += 1
                sIndex = sTempIndex
            else:
                return False

        while pIndex < len(p) and p[pIndex] == '*':
            pIndex += 1

        return pIndex == len(p)